### DISCUSSION TITLE
Kargo integration with Holos

### DISCUSSION THREAD

**Discussion by jeffmccune** [akuity/kargo#2995] [General]
I'm the maintainer of [Holos](https://holos.run/docs/v1alpha5/tutorial/overview/) which is an implementation of the [rendered manifest pattern](https://akuity.io/blog/the-rendered-manifests-pattern) in a Go command line tool with a [CUE](https://cuelang.org/) platform integration layer.

```mermaid
---
title: Rendering Overview
---
graph LR
    Platform[<a href="https://holos.run/docs/v1alpha5/api/author/#Platform">Platform</a>]
    Component[<a href="https://holos.run/docs/v1alpha5/api/author/#ComponentConfig">Components</a>]

    Helm[<a href="https://holos.run/docs/v1alpha5/api/author/#Helm">Helm</a>]
    Kustomize[<a href="https://holos.run/docs/v1alpha5/api/author/#Kustomize">Kustomize</a>]
    Kubernetes[<a href="https://holos.run/docs/v1alpha5/api/author/#Kubernetes">Kubernetes</a>]

    BuildPlan[<a href="https://holos.run/docs/v1alpha5/api/core/#BuildPlan">BuildPlan</a>]

    ResourcesArtifact[<a href="https://holos.run/docs/v1alpha5/api/core/#Artifact">Resources<br/>Artifact</a>]
    GitOpsArtifact[<a href="https://holos.run/docs/v1alpha5/api/core/#Artifact">GitOps<br/>Artifact</a>]

    Generators[<a href="https://holos.run/docs/v1alpha5/api/core/#Generator">Generators</a>]
    Transformers[<a href="https://holos.run/docs/v1alpha5/api/core/#Transformer">Transformers</a>]
    Validators[<a href="https://holos.run/docs/v1alpha5/api/core/#Validator">Validators</a>]
    Files[Manifest<br/>Files]

    Platform --> Component
    Component --> Helm --> BuildPlan
    Component --> Kubernetes --> BuildPlan
    Component --> Kustomize --> BuildPlan

    BuildPlan --> ResourcesArtifact --> Generators
    BuildPlan --> GitOpsArtifact --> Generators

    Generators --> Transformers --> Validators --> Files
```

Currently we handle promotion outside of the tool in a GitHub actions workflow.  On new container images the action workflow writes the new image tag to a JSON file then executes `holos render platform` and commits the updated platform configuration for ArgoCD to deploy.  I'm planning to integrate Holos with Kargo to take advantage of more advanced and well defined promotion steps across environments and regions.  This will likely take the form of a new `holos` subcommand to update the configuration data as one step, similar to the `kustomize edit` step Cargo has today, then a `holos render platform` step to produce the fully rendered manifests to commit back to GitOps.

This integration depends on custom/third-party promotion steps planned for v1.3.0.  I've been searching for design docs / issues related to this feature but haven't found any yet.  I'd like to connect with the engineer(s) working on the feature to hopefully act as an external design partner providing use cases for the integration.  Beyond design, I'd like to help implement the feature you design if that's something you're open to.  What's the best way to connect and help out with this feature?  

Thanks for making great tooling!

**Comment by jeffmccune** [comment]
Kent mentioned in 1250 discussing this in #2996

**Reply by joebowbeer** [reply]
#1250 adding link...

**Comment by joebowbeer** [comment]
@jeffmccune Adding a description of your current approach below.

> I think the best way to integrate Kargo into the rest of the platform is to minimize what Kargo does.  Reduce it down to simply bumping a version number in a yaml or json file in Git.  Then push the change and let CI run `holos render platform` which will pick up the version bump and do whatever you want it to do with the new version.
> 
> The rendered YAML would then be committed back to Git and pushed in the same CI pipeline.
> 
> Then, ArgoCD would pick up the fully rendered manifests and deploy them.

See https://github.com/holos-run/kargo-demo
